lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1111 @@
+			}
+		}
+	};
+	(LOCAL_OR_REMOTE_CHANGE, $to_local_transition: ident, $to_remote_transition: ident, $msg: ty) => {
+		define_state_machine_transitions!($to_local_transition, $msg, [
+			FROM RemoteChange, TO LocalChange,
+			FROM RemoteTxComplete, TO LocalChange
+		]);
+		define_state_machine_transitions!($to_remote_transition, $msg, [
+			FROM LocalChange, TO RemoteChange,
+			FROM LocalTxComplete, TO RemoteChange
+		]);
+	};
+}
+
+impl StateMachine {
+	fn new(feerate_sat_per_kw: u32, is_initiator: bool, tx_locktime: PackedLockTime) -> Self {
+		let context = NegotiationContext {
+			tx_locktime,
+			holder_is_initiator: is_initiator,
+			received_tx_add_input_count: 0,
+			received_tx_add_output_count: 0,
+			inputs: HashMap::new(),
+			prevtx_outpoints: HashSet::new(),
+			outputs: HashMap::new(),
+			feerate_sat_per_kw,
+		};
+		if is_initiator {
+			Self::RemoteChange(RemoteChange(context))
+		} else {
+			Self::LocalChange(LocalChange(context))
+		}
+	}
+
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_add_input, remote_tx_add_input, &msgs::TxAddInput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_add_output, remote_tx_add_output, &msgs::TxAddOutput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_remove_input, remote_tx_remove_input, &msgs::TxRemoveInput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_remove_output, remote_tx_remove_output, &msgs::TxRemoveOutput
+	);
+	define_state_machine_transitions!(local_tx_complete, &msgs::TxComplete, [
+		FROM RemoteChange, TO LocalTxComplete,
+		FROM RemoteTxComplete, TO NegotiationComplete
+	]);
+	define_state_machine_transitions!(remote_tx_complete, &msgs::TxComplete, [
+		FROM LocalChange, TO RemoteTxComplete,
+		FROM LocalTxComplete, TO NegotiationComplete
+	]);
+}
+
+pub struct InteractiveTxConstructor {
+	state_machine: StateMachine,
+	channel_id: ChannelId,
+	inputs_to_contribute: Vec<(SerialId, TxIn, Transaction)>,
+	outputs_to_contribute: Vec<(SerialId, TxOut)>,
+}
+
+pub enum InteractiveTxMessageSend {
+	TxAddInput(msgs::TxAddInput),
+	TxAddOutput(msgs::TxAddOutput),
+	TxComplete(msgs::TxComplete),
+}
+
+macro_rules! do_state_transition {
+	($self: ident, $transition: ident, $msg: expr) => {{
+		let state_machine = core::mem::take(&mut $self.state_machine);
+		$self.state_machine = state_machine.$transition($msg);
+		match &$self.state_machine {
+			StateMachine::NegotiationAborted(state) => Err(state.0.clone()),
+			_ => Ok(()),
+		}
+	}};
+}
+
+fn generate_local_serial_id<ES: Deref>(entropy_source: &ES, is_initiator: bool) -> SerialId where ES::Target: EntropySource {
+	let rand_bytes = entropy_source.get_secure_random_bytes();
+	let mut serial_id_bytes = [0u8; 8];
+	serial_id_bytes.copy_from_slice(&rand_bytes[..8]);
+	let mut serial_id = u64::from_be_bytes(serial_id_bytes);
+	if serial_id.is_valid_for_initiator() != is_initiator {
+		serial_id ^= 1;
```

Comment:
worth highlighting we lose one bit of entropy here, not sure if it matters if our seed is sufficiently random (?)

Reply:
It shouldn't matter really. In a multiparty scenario, this is what ends up happening anyway when a node forwards a tx_add_input/output message.